### ISSUE TITLE
Feature Request: structs can define their own update commands

### ISSUE THREAD

**Issue by rattrayalex** [gcanti/tcomb#247]
Currently you can add commands to the global `t.update` function. This is useful but could get bloated and messy if you add too many commands, many of which would probably be intended for use only in certain structs.

It would be nice to be able to do something like this: 

```js
var Office = t.struct({ address: t.Str, timesMoved: t.Num }, {
  defaultProps: { timesMoved: 0 },
  commands: {
    $moveTo: function(newAddress, oldOffice) {
       return { address: newAddress, timesMoved: oldOffice.timesMoved + 1 }
    }
  }
});

var office = Office({ address: '3rd street' });
var newOffice = Office.update(office, { $moveTo: '4th street' })
```

It's a bit of a silly example. In this case I think it could be done by `Office.prototype.moveOffice` or something, which could wrap `update`. But in more complex cases it would be nice to have custom commands that are scoped to that struct only. 

Let me know if you are interested and I can try to submit a PR. It seems fairly straightforward.

**Comment by gcanti** [comment]
> it could be done by `Office.prototype.moveOffice`

Hello @rattrayalex, 

Defining a static function or a method seems indeed the proper way

**Comment by rattrayalex** [comment]
Hmm, why allow users to add commands to the global `update` function but not to individual structs?

**Comment by gcanti** [comment]
Because commands are intended 

1) to be general
2) to be used in parallel

```js
import t from 'tcomb'

const x = {
  a: 1,
  b: [2, 3]
}

const y = t.update(x, {
  a: { $set: 4 },
  b: { $swap: { from: 0, to: 1 } }
})
```

Here `t.update` performs 2 operations with only 1 call

**Comment by rattrayalex** [comment]
> 2) to be used in parallel

Yes, exactly. Sorry my example wasn't a very good one, that did not show this. But if I wanted to `$moveTo` at the same time as doing some other operation on the office – say, `{ numEmployees: { $set: 500 } }`, for example – then I would not be able to by simply using a static function or method for `$moveTo`. 

> 1) to be general

Hmm, why is that? Unique datatypes often have unique transformations.

**Comment by gcanti** [comment]
> Unique datatypes often have unique transformations

Sure, but IMO it's better to define the smallest number of general and composable commands to build them all. For example the `$moveTo` command can be implemented as an `$apply` 

```js
const moveTo = (address) => ({ timesMoved }) => ({ address, timesMoved: timesMoved + 1 })
const office = new Office({ address: '3rd street' })
const newOffice = Office.update(office, { $apply: moveTo('4th street') })
```

**Comment by rattrayalex** [comment]
Hmm, fair enough. Thanks for taking the time to explain!
